packages/itwinui-react/src/utils/hooks/useControlledState.ts

```diff
@@ -27,14 +30,40 @@ export const useControlledState = <T>(
     () => (controlledState !== undefined ? controlledState : uncontrolledState),
     [controlledState, uncontrolledState],
   );
+  const oldState = React.useRef<T>(state);
 
   const setState = React.useCallback(
     (value) => {
+      const newValue = (() => {
+        if (typeof value === 'function') {
+          return (value as (prevState: T) => T)(state);
+        }
+        return value;
+      })();
+
+      // Called only when the value *changes* (uncontrolled mode) or should *change* (controlled mode).
+      if (newValue === oldState.current) {
+        return;
+      }
+
+      // If in controlled mode, defer updating oldState to the sync effect.
+      // Else, update oldState here since uncontrolledState is guaranteed to be set to value.
+      if (controlledState == null) {
+        oldState.current = newValue;
+      }
+
       setUncontrolledState(value);
       setControlledState?.(value);
     },
-    [setControlledState, setUncontrolledState],
+    [controlledState, setControlledState, state],
   ) as React.Dispatch<React.SetStateAction<T>>;
 
+  // If in controlled mode, sync oldState with controlledState
+  React.useEffect(() => {
+    if (controlledState != null) {
+      oldState.current = controlledState;
+    }
+  }, [controlledState]);
```

Comment:
Thinking about alternative approaches:
- Would `useLatestRef(state)` do the same thing?
- Will the ["Storing information from previous renders"](https://react.dev/reference/react/useState#storing-information-from-previous-renders) pattern work?